p2p/host/autorelay/autorelay.go

```diff
@@ -2,532 +2,142 @@ package autorelay
-				// Make sure we only start it once.
-				if atomic.CompareAndSwapInt32(&ar.findRelaysRunning, 0, 1) {
-					go func() {
-						defer atomic.StoreInt32(&ar.findRelaysRunning, 0)
-						ar.findRelays(ctx)
-					}()
+			switch evt.Reachability {
+			case network.ReachabilityPrivate, network.ReachabilityUnknown:
+				if err := r.relayFinder.Start(); err != nil {
+					log.Error("failed to start relay finder")
 				}
+			case network.ReachabilityPublic:
+				r.relayFinder.Stop()
 			}
-
-			ar.mx.Lock()
-			// if our reachability changed
-			if ar.status != evt.Reachability && evt.Reachability != network.ReachabilityUnknown {
-				push = true
-			}
-			ar.status = evt.Reachability
-			ar.mx.Unlock()
-		case <-ar.relayFound:
-			push = true
-		case now := <-ticker.C:
-			push = ar.refreshReservations(ctx, now)
-		case <-ctx.Done():
-			return
-		}
-
-		if push {
-			ar.mx.Lock()
-			ar.cachedAddrs = nil
-			ar.mx.Unlock()
-			ar.host.SignalAddressChange()
-		}
-	}
-}
-
-func (ar *AutoRelay) refreshReservations(ctx context.Context, now time.Time) bool {
-	ar.mx.Lock()
-	if ar.status == network.ReachabilityPublic {
-		// we are public, forget about the relays, unprotect peers
-		for p := range ar.relays {
-			ar.host.ConnManager().Unprotect(p, autorelayTag)
-			delete(ar.relays, p)
-		}
-
-		ar.mx.Unlock()
-		return true
-	}
-
-	if len(ar.relays) == 0 {
-		ar.mx.Unlock()
-		return false
-	}
-
-	// find reservations about to expire and refresh them in parallel
-	g := new(errgroup.Group)
-	for p, rsvp := range ar.relays {
-		if rsvp == nil {
-			// this is a circuitv1 relay, there is no reservation
-			continue
-		}
-		if now.Add(rsvpExpirationSlack).Before(rsvp.Expiration) {
-			continue
-		}
-
-		p := p
-		g.Go(func() error { return ar.refreshRelayReservation(ctx, p) })
-	}
-	ar.mx.Unlock()
-
-	err := g.Wait()
-	return err != nil
-}
-
-func (ar *AutoRelay) refreshRelayReservation(ctx context.Context, p peer.ID) error {
-	rsvp, err := circuitv2.Reserve(ctx, ar.host, peer.AddrInfo{ID: p})
-
-	ar.mx.Lock()
-	defer ar.mx.Unlock()
-
-	if err != nil {
-		log.Debugf("failed to refresh relay slot reservation with %s: %s", p, err)
-
-		delete(ar.relays, p)
-		// unprotect the connection
-		ar.host.ConnManager().Unprotect(p, autorelayTag)
-	} else {
-		log.Debugf("refreshed relay slot reservation with %s", p)
-		ar.relays[p] = rsvp
-	}
-
-	return err
-}
-
-func (ar *AutoRelay) findRelays(ctx context.Context) {
-	timer := time.NewTimer(30 * time.Second)
-	defer timer.Stop()
-	for retry := 0; retry < 5; retry++ {
-		if retry > 0 {
-			log.Debug("no relays connected; retrying in 30s")
+			r.mx.Lock()
+			r.status = evt.Reachability
+			r.mx.Unlock()
+		case pi := <-peerChan:
 			select {
-			case <-timer.C:
-			case <-ctx.Done():
-				return
+			case r.peerChanOut <- pi: // if there's space in the channel, great
+			default:
+				// no space left in the channel. Drop the oldest entry.
+				<-r.peerChanOut
+				r.peerChanOut <- pi
```

Comment:
isnt this racey with the consumer? I think we should also select here.

Reply:
Good catch!